File: spring-web-java/src/main/java/spring/web/java/application/service/RoleService.java
package spring.web.java.application.service;

import java.util.List;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import lombok.RequiredArgsConstructor;
import spring.web.java.domain.model.entity.Permission;
import spring.web.java.domain.model.entity.Role;
import spring.web.java.domain.model.entity.RolePermission;
import spring.web.java.domain.repository.PermissionRepository;
import spring.web.java.domain.repository.RoleRepository;
import spring.web.java.presentation.dto.request.RoleSaveRequest;
import spring.web.java.presentation.dto.response.RoleResponse;
import spring.web.java.presentation.exception.EntityNotFoundException;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class RoleService {

	private final RoleRepository roleRepository;
	private final PermissionRepository permissionRepository;

	@Transactional
	public RoleResponse saveRole(RoleSaveRequest roleSaveRequest) {
		List<RolePermission> rolePermissions = roleSaveRequest.permissionIds().stream()
			.map(id -> {
				Permission permission = permissionRepository.findById(id)
					.orElseThrow(() -> new EntityNotFoundException(Permission.class, id));

				return RolePermission.create(permission);
			})
			.toList();

		Role role = roleRepository.save(Role.create(roleSaveRequest.name(), rolePermissions));

		return new RoleResponse(
			role,
			role.getRolePermissions().stream()
				.map(RolePermission::getPermission)
				.toList()
		);
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/application/service/RoleService.kt
package spring.web.kotlin.application.service

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import spring.web.kotlin.domain.model.entity.Permission
import spring.web.kotlin.domain.model.entity.Role
import spring.web.kotlin.domain.model.entity.RolePermission
import spring.web.kotlin.domain.repository.PermissionRepository
import spring.web.kotlin.domain.repository.RoleRepository
import spring.web.kotlin.presentation.dto.request.RoleSaveRequest
import spring.web.kotlin.presentation.dto.response.RoleResponse
import spring.web.kotlin.presentation.exception.EntityNotFoundException

@Service
@Transactional(readOnly = true)
class RoleService(
    private val roleRepository: RoleRepository,
    private val permissionRepository: PermissionRepository
) {
    @Transactional
    fun saveRole(roleSaveRequest: RoleSaveRequest): RoleResponse {
        val rolePermissions = roleSaveRequest.permissionIds.map {
            val permission = permissionRepository.findByIdOrNull(it)
                ?: throw EntityNotFoundException(Permission::class.java, it)

            RolePermission.create(permission)
        }

        val role = roleRepository.save(Role.create(roleSaveRequest.name, rolePermissions))

        return RoleResponse(role, role.rolePermissions.mapNotNull { it.permission })
    }
}